plugins/modules/ec2_spot_instance.py

```diff
@@ -0,0 +1,475 @@
+        module.fail_json_aws(e, msg='Error while cancelling the spot instance request')
+
+
+def main():
+    argument_spec = dict(
+        zone_group=dict(type='str', default='  '),
+        client_token=dict(type='str', default='  '),
+        count=dict(type='int', default=1),
+        interruption=dict(type='str', default="terminate"),
+        launch_group=dict(type='str', default='  '),
+        launch_specification=dict(type='dict', default=dict()),
+        request_type=dict(default='one-time', choices=["one-time", "persistent"]),
+        state=dict(default='present', choices=['present', 'absent', 'running', 'restarted', 'stopped']),
+        device_name=dict(type='str'),
+        virtual_name=dict(type='str'),
+        ebs=dict(type='dict'),
+        no_device=dict(type='str'),
+        block_device_mappings=dict(type='list',elements='dict'),
+        network_interfaces=dict(type='list', elements='dict'),
+        associate_public_ip_address=dict(type='bool'),
+        delete_on_termination=dict(type='bool'),
+        description=dict(type='str'),
+        decive_index=dict(type='int'),
+        groups=dict(type='list'),
+        ipv6_address_count=dict(type='int'),
+        ipv6_addresses=dict(type=list, elements='dict'),
+        network_interface_id=dict(type='str'),
+        private_ip_address=dict(type='str'),
+        private_ip_addresses=dict(type='list', elements='dict'),
+        secondary_private_ip_address_count=dict(type='int'),
+        subnet_id=dict(type='str'),
+        associate_carrier_ip_address=dict(type='boolean'),
+        interface_type=dict(type='str', choices=['interface', 'efa']),
+        network_card_index=dict(type='int'),
+        ipv4_prefixes=dict(type='list', elements='dict'),
+        ipv4_prefix_count=dict(type='int'),
+        ipv6_prefixes=dict(type='list', elements='dict'),
+        ipv6_prefix_count=dict(type='int'),
+        spot_price=dict(type='str', default=''),
+        spot_type=dict(default='one-time', choices=["one-time", "persistent"]),
+        tags=dict(type='list', default=[]),
+        # valid_from=dict(type='datetime', default=datetime.datetime.now()),
+        # valid_until=dict(type='datetime', default=(datetime.datetime.now() + datetime.timedelta(minutes=60))
+        spot_instance_request_ids=dict(type='list', elements='str', default=[]),
+        placement=dict(type='dict'),
+        monitoring=dict(type='dict'),
+    )
+    module = AnsibleAWSModule(
+        argument_spec=argument_spec,
+        check_boto3=False,
+        mutually_exclusive=[
+            ['block_duration', 'zone_group'],
+            ['block_duration', 'launch_group'],
+        ],
+        supports_check_mode=True
+    )
+
+    connection = module.client('ec2', AWSRetry.jittered_backoff())
+
+    state = module.params['state']
+
+    if state == 'present':
+        response = request_spot_instances(module, connection)
+        changed = True
```

Comment:
What does the API do if a request that matches all the parameters already exists?  Does it recreate it or is there no change?

Reply:
I have been running the same playbook for testing. It creates a new request each time since even if the parameters are the same, Boto3 considers it as a new request for a spot instance with the same configuration.

Reply:
In this case we should handle idempotency inside the module by checking for existing requests and comparing them to the specification then.  Let's get the current functionality sorted out and then we can work on this.

Reply:
Sure!